README.md
# NAME

Courriel - High level email parsing and manipulation

# VERSION

version 0.43

# SYNOPSIS

    my $email = Courriel->parse( text => $raw_email );

    print $email->subject;

    print $_->address for $email->participants;

    print $email->datetime->year;

    if ( my $part = $email->plain_body_part ) {
        print $part->content;
    }

# DESCRIPTION

This class exists to provide a high level API for working with emails,
particular for processing incoming email. It is primarily a wrapper around the
other classes in the Courriel distro, especially [Courriel::Headers](https://metacpan.org/pod/Courriel::Headers),
[Courriel::Part::Single](https://metacpan.org/pod/Courriel::Part::Single), and [Courriel::Part::Multipart](https://metacpan.org/pod/Courriel::Part::Multipart). If you need lower
level information about an email, it should be available from one of these
classes.

# API

This class provides the following methods:

## Courriel->parse( text => $raw\_email, is\_character => 0|1 )

This parses the given text and returns a new Courriel object. The text can be
provided as a string or a reference to a string.

If you pass a reference, then the scalar underlying the reference _will_ be
modified, so don't pass in something you don't want modified.

By default, Courriel expects that content passed in text is binary data. This
means that it has not been decoded into utf-8 with `Encode::decode()` or by
using a `:encoding(UTF-8)` IO layer.

In practice, this doesn't matter for most emails, since they either contain
only ASCII data or they actually do contain binary (non-character)
data. However, if an email is using the 8bit Content-Transfer-Encoding, then
this does matter.

If the email has already been decoded, you must set `is_character` to a true
value.

It's safest to simply pass binary data to Courriel and let it handle decoding
internally.

## $email->parts()

Returns an array (not a reference) of the parts this email contains.

## $email->part\_count()

Returns the number of parts this email contains.

## $email->is\_multipart()

Returns true if the top-level part is a multipart part, false otherwise.

## $email->top\_level\_part()

Returns the actual top level part for the object. You're probably better off
just calling `$email->parts()` most of the time, since when the email is
multipart, the top level part is just a container.

## $email->subject()

Returns the email's Subject header value, or `undef` if it doesn't have one.

## $email->datetime()

Returns a [DateTime](https://metacpan.org/pod/DateTime) object for the email. The DateTime object is always in
the "UTC" time zone.

This uses the Date header by default one. Otherwise it looks at the date in
each Received header, and then it looks for a Resent-Date header. If none of
these exists, it just returns `DateTime->now()`.

## $email->from()

This returns a single [Email::Address](https://metacpan.org/pod/Email::Address) object based on the From header of the
email. If the email has no From header or if the From header is broken, it
returns `undef`.

## $email->participants()

This returns a list of [Email::Address](https://metacpan.org/pod/Email::Address) objects, one for each unique
participant in the email. This includes any address in the From, To, or CC
headers.

Just like with the From header, broken addresses will not be included.

## $email->recipients()

This returns a list of [Email::Address](https://metacpan.org/pod/Email::Address) objects, one for each unique
recipient in the email. This includes any address in the To or CC headers.

Just like with the From header, broken addresses will not be included.

## $email->to()

This returns a list of [Email::Address](https://metacpan.org/pod/Email::Address) objects, one for each unique
address in the To header.

Just like with the From header, broken addresses will not be included.

## $email->cc()

This returns a list of [Email::Address](https://metacpan.org/pod/Email::Address) objects, one for each unique
address in the CC header.

Just like with the From header, broken addresses will not be included.

## $email->plain\_body\_part()

This returns the first [Courriel::Part::Single](https://metacpan.org/pod/Courriel::Part::Single) object in the email with a
mime type of "text/plain" and an inline disposition, if one exists.

## $email->html\_body\_part()

This returns the first [Courriel::Part::Single](https://metacpan.org/pod/Courriel::Part::Single) object in the email with a
mime type of "text/html" and an inline disposition, if one exists.

## $email->clone\_without\_attachments()

Returns a new Courriel object that only contains inline parts from the
original email, effectively removing all attachments.

## $email->first\_part\_matching( sub { ... } )

Given a subroutine reference, this method calls that subroutine for each part
in the email, in a depth-first search.

The subroutine receives the part as its only argument. If it returns true,
this method returns that part.

## $email->all\_parts\_matching( sub { ... } )

Given a subroutine reference, this method calls that subroutine for each part
in the email, in a depth-first search.

The subroutine receives the part as its only argument. If it returns true,
this method includes that part.

This method returns all of the parts that match the subroutine.

## $email->content\_type()

Returns the [Courriel::Header::ContentType](https://metacpan.org/pod/Courriel::Header::ContentType) object associated with the email.

## $email->headers()

Returns the [Courriel::Headers](https://metacpan.org/pod/Courriel::Headers) object for this email.

## $email->stream\_to( output => $output )

This method will send the stringified email to the specified output. The
output can be a subroutine reference, a filehandle, or an object with a
`print()` method. The output may be sent as a single string, as a list of
strings, or via multiple calls to the output.

For large emails, streaming can be much more memory efficient than generating
a single string in memory.

## $email->as\_string()

Returns the email as a string, along with its headers. Lines will be
terminated with "\\r\\n".

# ROBUSTNESS PRINCIPLE

Courriel aims to respect the common Internet robustness principle (aka
Postel's law). Courriel is conservative in the output it generates, and
liberal in what it accepts.

When parsing, the goal is to never die and always return as much information
as possible. Any input that causes the `Courriel->parse()` to die means
there's a bug in the parser. Please report these bugs.

Conversely, Courriel aims to respect all relevant RFCs in its output, except
when it preserves the original data in a parsed email. If you're using
[Courriel::Builder](https://metacpan.org/pod/Courriel::Builder) to create emails from scratch, any output that isn't
RFC-compliant is a bug.

# FUTURE PLANS

This release is still rough, and I have some plans for additional features:

## More methods for walking all parts

Some more methods for walking/collecting multiple parts would be useful.

## More?

Stay tuned for details.

# WHY DID I WRITE THIS MODULE?

There a lot of email modules/distros on CPAN. Why didn't I use/fix one of them?

- [Mail::Box](https://metacpan.org/pod/Mail::Box)

    This one probably does everything this module does and more, but it's really,
    really big and complicated, forcing the end user to make a lot of choices just
    to get started. If you need it, it's great, but I generally find it to be too
    much module for me.

- [Email::Simple](https://metacpan.org/pod/Email::Simple) and [Email::MIME](https://metacpan.org/pod/Email::MIME)

    These are surprisingly **not** simple. They suffer from a problematic API (too
    high level in some spots, too low in others), and a poor separation of
    concerns. I've hacked on these enough to know that I can never make them do
    what I want.

- Everything Else

    There's a lot of other email modules on CPAN, but none of them really seem any
    better than the ones mentioned above.

# CREDITS

This module rips some chunks of code from a few other places, notably several
of the Email suite modules.

# DONATIONS

If you'd like to thank me for the work I've done on this module, please
consider making a "donation" to me via PayPal. I spend a lot of free time
creating free software, and would appreciate any support you'd care to offer.

Please note that **I am not suggesting that you must do this** in order for me
to continue working on this particular software. I will continue to do so,
inasmuch as I have in the past, for as long as it interests me.

Similarly, a donation made in this way will probably not make me work on this
software much more, unless I get so many donations that I can consider working
on free software full time, which seems unlikely at best.

To donate, log into PayPal and send money to autarch@urth.org or use the
button on this page: [http://www.urth.org/~autarch/fs-donation.html](http://www.urth.org/~autarch/fs-donation.html)

# BUGS

Please report any bugs or feature requests to `bug-courriel@rt.cpan.org`, or
through the web interface at [http://rt.cpan.org](http://rt.cpan.org).  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.

Bugs may be submitted through [the RT bug tracker](http://rt.cpan.org/Public/Dist/Display.html?Name=Courriel)
(or [bug-courriel@rt.cpan.org](mailto:bug-courriel@rt.cpan.org)).

I am also usually active on IRC as 'autarch' on `irc://irc.perl.org`.

# DONATIONS

If you'd like to thank me for the work I've done on this module, please
consider making a "donation" to me via PayPal. I spend a lot of free time
creating free software, and would appreciate any support you'd care to offer.

Please note that **I am not suggesting that you must do this** in order for me
to continue working on this particular software. I will continue to do so,
inasmuch as I have in the past, for as long as it interests me.

Similarly, a donation made in this way will probably not make me work on this
software much more, unless I get so many donations that I can consider working
on free software full time (let's all have a chuckle at that together).

To donate, log into PayPal and send money to autarch@urth.org, or use the
button at [http://www.urth.org/~autarch/fs-donation.html](http://www.urth.org/~autarch/fs-donation.html).

# AUTHOR

Dave Rolsky <autarch@urth.org>

# CONTRIBUTORS

- Gregory Oschwald <goschwald@maxmind.com>
- Ricardo Signes <rjbs@users.noreply.github.com>
- Zbigniew Łukasiak <zzbbyy@gmail.com>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2016 by Dave Rolsky.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)


Changes
{{$NEXT}}

0.49     2021-08-08

- Fixed handling of charsets in multi-section header attributes. Previously
  this would assume that the charset was present in every section, instead of
  just the first. Fixed by Vitaly Gimly. GH #10. PR #11.

- Fixed handling of multi-section header attributes which encoded a UTF-8 (or
  other multi-byte charset) value with the value broken mid-byte between
  sections. Previously these would be end up in a corrupted form. Reported by
  Vitaly Gimly. GH #10.

- Fixed parsing of MIME encoded words in a header when there was no space at
  the start of the value. Previously this would leave the first MIME encoded
  chunk as-is instead of decoding it. Fixed by Vitaly Gimly. GH #10. PR #11.


0.48     2019-04-21

- Replaced File::Slurp::Tiny with File::Slurper. The former has been
  deprecated by its author for some time. Fixes GH #9. Reported by Greg
  Oschwald.


0.47     2018-05-20

- Switched to using GitHub for issue tracking.


0.46     2018-05-13

- Replaced all use of Email::Address with Email::Address::XS, as the former is
  deprecated. The latter is a drop-in replacement API-wise, except if you are
  checking `$address->isa('Email::Address')`. Reported by Greg Oschwald. GH
  #5.


0.45     2017-12-09

- Fix tests that fail with Params::ValidationCompiler 0.26 and require that
  version.


0.44     2016-12-05

- The changes to use Params::ValidationCompiler in 0.43 were not quite
  correct, and since Params::ValidationCompiler 0.20 started doing more
  validation of calls to validation_for(), this cause an exception. Reported
  by Todd Bruner. RT #119137.


0.43     2016-11-13

- Replaced MooseX::Params::Validate with Params::ValidationCompiler.


0.42     2016-07-04

- Require the latest version of DateTime::Format::Mail in order to avoid
  warnings from DateTime 1.32+ that will cause some tests to fail. Reported by
  Olaf Alders. RT #115868.


0.41     2016-03-05

- Fixed header folding for headers that contain non-true values like
  '0'. Based on a patch for Email::Simple from
  pali. https://github.com/rjbs/Email-Simple/pull/11.


0.40     2016-01-30

- Fixed the stringification of some headers with regards to encoding. The
  Received header should never be encoded (and thus should never contain
  non-ASCII).

- Fixed headers which contain email addresses (To, From, CC) so that email
  addresses are properly encoded (I think). The phrase and comment portions of
  the address can contain encoded-words, but the address itself should never
  be encoded. Note that per RFC6531 addresses may contain UTF-8, but the MTA
  must support the SMTPUTF8 extension in order for the mail to be
  delivered. As of this release, SMTPUTF8 support does not appear to be very
  widespread, so I would strongly discourage the use of UTF-8 in email
  addresses!


0.39     2015-08-04

- Changed Email::Abstract::Courriel's ->get_body() method to return the entire
  email as a string, making it equivalent to ->as_string(). This is confusing,
  but looking at the few instances of code in Email::Transport that call
  ->get_body(), it seems like that's what this method is expected to
  do. Reported by Greg Oschwald. RT #106176.


0.38     2015-08-04

- Somewhat screwy release now deleted from PAUSE.


0.37     2015-02-26

- Fixed a test failure when MooseX::Params::Validate 0.21+ is
  installed. Reported by Slaven Rezic. RT #102382.


0.36     2014-08-15

- Trying to build an email with a header that started with a space caused the
  header value to be dropped entirely from the generated email. Patch by Greg
  Oschwald. PR #2.


0.35     2014-05-12

- Replaced all uses of File::Slurp with File::Slurp::Tiny. See RT #95479 for
  the initial bug report and discussion on why File::Slurp should not be used.


0.34     2014-05-02

- More test fixes - relax the date parsing more.


0.33     2014-04-19

- Fixed tests so they pass with the most recent DateTime::Format::Mail (0.400)
  as well as older versions.


0.32     2014-04-19

- Courriel parsing is now more liberal and will accept an email with mixed
  line endings (some newline, some CRLF, etc.). Apparently this happens in the
  wild (of course it does). Reported by Joseph Crotty. RT #94837.


0.31     2013-04-03

[MISC]

- Removed the "this is very alpha" warning from the docs. It's really not very
  alpha any more.


0.30     2012-08-20

[API CHANGES]

- Courriel::Builder no longer exports a bcc() subroutine. Creating Bcc headers
  in emails does not make any sense whatsoever. See
  https://metacpan.org/module/Email::Sender::Manual::QuickStart#Hey-wheres-my-Bcc-support-
  for an explanation of why there is no such thing as a Bcc header.


0.29     2012-03-07

[BUG FIXES]

- Courriel now allows an email to have an empty subject when parsing, as does
  Courriel::Builder. Based on a patch from Zbigniew Łukasiak.


0.28     2012-02-01

[BUG FIXES]

- Header attribute names need to be looked up case-insensitively, according to
  the spec. The original casing is preserved in the Courriel::HeaderAttribute
  object's name(), method, but looking up an attribute by name on a
  Courriel::Header object now accepts any casing for the name.


0.27     2011-09-24

[ENHANCEMENTS]

- All objects which can stringify now support a ->stream_to API as well. This
  lets you stream their string output to a sub ref, filehandle, or an object
  with a ->print() method. If you are dealing with large emails, this API may
  be more efficient than generating a single string in memory.

[API CHANGES]

- Renamed the Header->as_header_string and HeaderAttribute->as_header_string
  methods to as_string, so that all objects have the same stringification API.


0.26     2011-09-19

[BUG FIXES]

- Add missing prereq for MooseX::Role::Parameterized. Reported by zby.


0.25     2011-09-15

[BUG FIXES]

- When a header value contained any non-ASCII content, Courriel would
  mistakenly MIME encode the whole value. In some cases, this is okay, but for

Makefile.PL
# This Makefile.PL for Courriel was generated by
# Dist::Zilla::Plugin::DROLSKY::MakeMaker 1.19
# and Dist::Zilla::Plugin::MakeMaker::Awesome 0.48.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.010000;
use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "High level email parsing and manipulation",
  "AUTHOR" => "Dave Rolsky <autarch\@urth.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Courriel",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.010000",
  "NAME" => "Courriel",
  "PREREQ_PM" => {
    "Carp" => 0,
    "DateTime" => 0,
    "DateTime::Format::Mail" => "0.403",
    "DateTime::Format::Natural" => 0,
    "Devel::PartialDump" => 0,
    "Email::Abstract::Plugin" => 0,
    "Email::Address::XS" => 0,
    "Email::MIME::Encodings" => 0,
    "Email::MessageID" => 0,
    "Encode" => 0,
    "Exporter" => 0,
    "File::Basename" => 0,
    "File::LibMagic" => 0,
    "File::Slurper" => 0,
    "List::AllUtils" => 0,
    "List::MoreUtils" => "0.28",
    "MIME::Base64" => 0,
    "MIME::QuotedPrint" => 0,
    "Moose" => 0,
    "Moose::Role" => 0,
    "MooseX::Role::Parameterized" => 0,
    "MooseX::StrictConstructor" => 0,
    "MooseX::Types" => 0,
    "MooseX::Types::Combine" => 0,
    "MooseX::Types::Common::String" => 0,
    "MooseX::Types::Moose" => 0,
    "Params::ValidationCompiler" => "0.26",
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "namespace::autoclean" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Sys::Hostname" => 0,
    "Test::Differences" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "Test::Requires" => 0,
    "Test::Warnings" => 0,
    "utf8" => 0
  },
  "VERSION" => "0.50",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

my %FallbackPrereqs = (
  "Carp" => 0,
  "DateTime" => 0,
  "DateTime::Format::Mail" => "0.403",
  "DateTime::Format::Natural" => 0,
  "Devel::PartialDump" => 0,
  "Email::Abstract::Plugin" => 0,
  "Email::Address::XS" => 0,
  "Email::MIME::Encodings" => 0,
  "Email::MessageID" => 0,
  "Encode" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Basename" => 0,
  "File::LibMagic" => 0,
  "File::Slurper" => 0,
  "File::Spec" => 0,
  "List::AllUtils" => 0,
  "List::MoreUtils" => "0.28",
  "MIME::Base64" => 0,
  "MIME::QuotedPrint" => 0,
  "Moose" => 0,
  "Moose::Role" => 0,
  "MooseX::Role::Parameterized" => 0,
  "MooseX::StrictConstructor" => 0,
  "MooseX::Types" => 0,
  "MooseX::Types::Combine" => 0,
  "MooseX::Types::Common::String" => 0,
  "MooseX::Types::Moose" => 0,
  "Params::ValidationCompiler" => "0.26",
  "Scalar::Util" => 0,
  "Sub::Exporter" => 0,
  "Sys::Hostname" => 0,
  "Test::Differences" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "Test::Requires" => 0,
  "Test::Warnings" => 0,
  "namespace::autoclean" => 0,
  "parent" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
